pkg/storage/mvcc.go

```diff
@@ -3906,6 +3935,215 @@ func MVCCGarbageCollect(
+		if rangeKey.StartKey.Compare(lastEndKey) < 0 {
+			return errors.Errorf("range keys in gc request should be non-overlapping: %s vs %s",
+				rangeKeys[i-1].String(), rangeKey.String())
+		}
+		lastEndKey = rangeKey.EndKey
+	}
+
+	var lastFragmentStart roachpb.Key
+	var pendingRanges []MVCCRangeKey
+	deletePendingFragments := func() error {
+		// Delete what was previously found.
+		if len(pendingRanges) > 0 {
+			for _, r := range pendingRanges {
+				if err := rw.ExperimentalClearMVCCRangeKey(r); err != nil {
+					return err
+				}
+			}
+			pendingRanges = pendingRanges[:0]
+		}
+		return nil
+	}
+
+	// Bound the iterator appropriately for the set of keys we'll be garbage
+	// collecting.
+	iter := rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
+		LowerBound: rangeKeys[0].StartKey,
+		UpperBound: rangeKeys[len(rangeKeys)-1].EndKey,
+		KeyTypes:   IterKeyTypePointsAndRanges,
+	})
+	defer iter.Close()
+
+	var currentKey roachpb.Key
+	for _, rangeKey := range rangeKeys {
+		iter.SeekGE(MVCCKey{Key: rangeKey.StartKey})
+
+		// Iterate within fragment till we find the end of it.
+		ok, err := iter.Valid()
+		if err != nil {
+			return err
+		}
+		if !ok {
+			if err := deletePendingFragments(); err != nil {
+				return err
+			}
+			break
+		}
+		hasP, hasR := iter.HasPointAndRange()
+
+		for ok { // Loop over gc'd range
+			if hasR {
+				span := iter.RangeBounds()
+				fragmentStart := span.Key
+				// Check if we landed on the new fragment set and need to find matching.
+				if !fragmentStart.Equal(lastFragmentStart) {
+					lastFragmentStart = append(lastFragmentStart[:0], fragmentStart...)
+					if err := deletePendingFragments(); err != nil {
+						return err
+					}
+					// Check if we moved past the end of GC range first.
+					if fragmentStart.Compare(rangeKey.EndKey) >= 0 {
+						break
+					}
+					// Find candidate fragments.
+					rangeKeyFragments := iter.RangeKeys()
+					idx := sort.Search(len(rangeKeyFragments), func(i int) bool {
+						return rangeKeyFragments[i].RangeKey.Timestamp.LessEq(rangeKey.Timestamp)
+					})
+					// Copy overlapping parts of range fragments below timestamp into pending.
+					if idx < len(rangeKeyFragments) {
+						stackStart := rangeKey.StartKey
+						if fragmentStart.Compare(rangeKey.StartKey) > 0 {
+							stackStart = fragmentStart.Clone()
+						}
+						stackEnd := rangeKey.EndKey
+						if span.EndKey.Compare(rangeKey.EndKey) < 0 {
+							stackEnd = span.EndKey.Clone()
+						}
+						for ; idx < len(rangeKeyFragments); idx++ {
+							pendingRanges = append(pendingRanges, MVCCRangeKey{
+								StartKey:  stackStart,
+								EndKey:    stackEnd,
+								Timestamp: rangeKeyFragments[idx].RangeKey.Timestamp,
+							})
+						}
+					}
+				}
+			} else {
+				if err := deletePendingFragments(); err != nil {
+					return err
+				}
+				lastFragmentStart = lastFragmentStart[:0]
+			}
+
+			// If point, we need to check anything below deleted tombstone.
```

Comment:
This is probably better handled by using an `MVCCIncrementalIterator`, which can use a TBI to optimize the scan. If we set one up with `EndTime = Timestamp` then we just need to do a seek and check if it finds anything.

It's a bit iffy that we keep reaching for `MVCCIncrementalIterator` every time we need to use a TBI, but it'll do for now since TBIs can be tricky.

I'm also a bit worried about the cost here, but let's revisit that later with a `TODO`.

Reply:
I had to create a separate mvcc incremental iterator as using one for everything will not allow checking ranges above GC threshold and this is needed to correctly recompute MVCCStats. But it looks much neater now and skips all the values in a correct case.